Replacement of D2K mod static shellmap with a scripted shellmap

Mix-martes86 opened the pull request:
Complete replacement of the fixed-screen shellmap currently in use for a customized shellmap, taking original RA mod shellmap as basis. WIP, terrain is not 100% finished, needs corrections, some buildings on sand instead of rock, and repetitive terrain patterns. Some scripting fixes done when rules for structures were modified. Commit 0cc324e. Reference (don't mind the music, used D2K's score track in script): https://www.youtube.com/watch?v=8Ioku0YT2F0

pchote committed:
Update terrain layer when the terrain data changes.

alexander-boll committed:
added mission nod4a

Polish nod04a a bit

abcdefg30 committed:
Convert the float health percentage to an int one

Mailaender committed:
Add menu plumbing for the new map editor.

pchote committed:
Add support for multiple logic objects on a widget.

alexander-boll committed:
added nod4b

Mailaender committed:
fix the cnc mpspawn and waypoint palettes

Robmaister committed:
Updated SharpFont to 3.0.0, adjusted font rendering code for new fixed-point types.

phrohdoh committed:
Add interface IPreventsEjectOnDeath.

penev92 committed:
Make buildings grant prerequisites explicitly

penev92 committed:
Add an upgrade rule for explicit prerequisite providing by buildings

penev92 committed:
Fix building prerequisites in RA

penev92 committed:
Fix building prerequisites in Cnc

penev92 committed:
Fix building prerequisites in D2k

Rydra committed:
Remove plumbing for trait unit tests.

phrohdoh committed:
Remove explicit private.
Remove unnecessary delegate.
Use extension method syntax.
Fix HealthInfo.NotifyAppliedDamage’s desc.
Remove unused using directives.
Remove explicit type declaration in Manifest.

RoosterDragon committed:
Refactored ScreenMap & improved perf of updates, removals and region lookups.

Reduce code duplication by extracting a common class to deal with spatial partitioning of actors, and use some (cached) delegates to reduce duplication further without affecting performance too much.

Speed up updates and removal of actors by caching their location so we only need to update or remove them from bins they are actually in (typically very few), compared to having to check every bin for removals which is much more work in comparison.

Speed up checking for actors inside a region by checking if items are located entirely within the bin they are located in. If so, we don't need to add them to the hash-set for de-duplication purposes which is fairly expensive.

sinf committed:
Added Hotkeys for support powers

sinf committed:
Fixed parsing of hotkeys with no name (such as tilde)

reaperrr committed:
Clean up infantry sequences

pchote committed:
Add Remove support to TypeDictionary.

reaperrr committed:
Clean up structure sequences

Now that we have tileset codes, those bitrotted file name definitions are no longer needed.

pchote committed:
Reorganize actor and smudge loading.

The actor and smudge definitions are now stored
as raw MiniYamlNodes in the map.  It is now the
responsibility of the consumers to parse these
into real objects.

pchote committed:
Account for damage state in ActorPreviews.

pchote committed:
Account for bibs in ActorPreviews.

pchote committed:
Adjust IRadarSignature to support per-cell colors.

penev92 committed:
Add "Enemy units detected" notifications to D2k

penev92 committed:
Add "Enemy detected" notification to RA

pchote committed:
Add support for wall preview neighbours.

Mailaender committed:
try to cache build dependency downloads

Mailaender committed:
fix GASPOT again

Mailaender committed:
try to cache the native DLL dependencies

pchote committed:
Fix crate actor previews.

Mailaender committed:
add 2 new blend modes: Multiplicative, DoubleMultiplicative

Mailaender committed:
setup the alpha image lighting for the light spot

sinf committed:
Add more TS notification sounds

penev92 committed:
Revert removal of EnemyWatcher and AnnounceOnSeen traits from RA and D2k

sinf committed:
Print GLSL info log if shader compilation fails

Mailaender committed:
add a global palette effect for day/night ambient

phrohdoh committed:
Implement DeployToUpgrade.
Expose some deploy-related cursors to yaml.

Mailaender committed:
add a looping sound trait

pchote committed:
Fix custom palettes in actor previews.

pchote committed:
Fix turret preview z-offsets.

pchote committed:
Import lighting parameters for TS maps.

reaperrr committed:
Fixed alphatst.shp

Mailaender committed:
add the Tiberian Sun crates

pchote committed:
Add methods to IOccupySpaceInfo.

pchote committed:
Add new map editor UI.

pchote committed:
Add the world components of the new editor.

Mailaender committed:
unhardcode and self-document render sequences

phrohdoh committed:
Add UnloadTerrainTypes restriction to Cargo.

pchote committed:
Replace OSX launcher with a native-code version.

- Binary files hosted externally.
- Fixes dialogs hijacking Finder.
- Fixes dock icon being lost when restarting.
- Adds openra:// url handling.
- Adds .orarep file handling.
- Minimum OS X version is now 10.6.

rneatherway committed:
Remove version information from NuGet folder names

This makes the names consistent and prevents changes in versions of
dependencies from breaking the build.

reaperrr committed:
De-hardcoded LaserZap animation sequence

reaperrr committed:
De-hardcoded GravityBomb idle and opening sequence

Avoid checking for sequence if Image is null.

RoosterDragon committed:
Added ConcurrentCache, a thread-safe Cache.

RoosterDragon committed:
Made FieldLoader thread-safe, and cache some expensive reflection calls.

RoosterDragon committed:
Load replays asynchronously and in parallel.

This prevents the UI blocking, and also speeds up loading time for getting all the replays displayed.

RoosterDragon committed:
Avoid ScrollPanelWidget drawing invisible child items.

If child items in a scroll panel will be outside the scissor area, then we can avoid drawing them at all. If a scroll panel has many items, this reduces to the draw cost closer to those visible in the panel, rather than costing for all the items.

penev92 committed:
Enable minimap toggling via the DisableShroud dev command

Mailaender committed:
install Linux app store data metadata

pchote committed:
Remove binaries from thirdparty/osx dir.

LunaticEdit committed:
EndOfStream first chance exceptions on server

The server class reads the memory stream in a forever loop until an
EndOfStreamException occurs. This causes repeated first chance
exceptions due to EndOfStreamException. This change verifies that the
memory stream's reader position is not past the length of the memory
stream.

Smith00101010 committed:
Added CnC: gdi05b

penev92 committed:
Change RA harv.shp

Mailaender committed:
fetch SDL2-CS from GitHub

Mailaender committed:
fetch Eluant from GitHub

Mailaender committed:
remove outdated documentation

Mailaender committed:
fix TS power down palette

abcdefg30 committed:
Adjust some selection bounds

No idea what regressed them.

abcdefg30 committed:
Fix HMEC walking animation

abcdefg30 committed:
Added BuildLimit: 1 to GHOST and HMEC

abcdefg30 committed:
Correct TRNSPORT prerequisites

pchote committed:
Include all the relevant palettes in the asset browser.

pchote committed:
Load tmp(ts) depth data.

Mailaender committed:
fix TS repair indicator palette

pchote committed:
Add a palette for previewing depth info in the asset browser.

penev92 committed:
Fix building prerequisites in TS

abcdefg30 committed:
Add descriptions for all buildable actors

abcdefg30 committed:
Remove bogus hotkey ("Hothey") definition

RoosterDragon committed:
Convert to 32bbp internally in FastCopyIntoSprite.

This avoids the need for callers to ensure the pixel format is correct, but ensures that the copying succeeds when the format is different.

Mailaender committed:
readonly and don't sync cosmetic traits

obrakmann committed:
Don't unnecessarily discard 24bpp map previews

Those images now get converted on-the-fly by the game.

obrakmann committed:
Convert RGB map preview image files to RGBA

Mailaender committed:
add the Hovercraft wake animation

pchote committed:
Don’t crash when asking for the offset of an invalid subcell.

pchote committed:
Fix orca’s flying sequences.

pchote committed:
Remove RenderSprites animation keys.

Mailaender committed:
add alpha channel to every possible blend mode

Mailaender committed:
add water wakes to APC

Mailaender committed:
separate with newlines for readability

Mailaender committed:
readonly

abcdefg30 committed:
Use OnPlayerDiscovered instead of OnEnteredFootprint

in Monster Tank Madness

abcdefg30 committed:
Add a new line at the end of descriptions for better readability

abcdefg30 committed:
Fix an issue in EnemyWatcher.cs

Mailaender committed:
fix transform offsets

Mailaender committed:
fix the artillery voxel lighting

pchote committed:
Don’t export certain Init types.

pchote committed:
Fix random country display.

pchote committed:
Separate terrain geometry visualization from DevMode.

Also adds a “terrainoverlay" console command.

abcdefg30 committed:
Remove the interval hack from gdi05a

pchote committed:
Fix a regression from #7987.

penev92 committed:
Reorganize PlayerReference loading to enable upgrade rules for them.

The PlayerReference definitions are now stored
as raw MiniYamlNodes in the map.

obrakmann committed:
Fix alignment issues in the score screen

Robmaister committed:
Removed zlib dependency and chose a specific version of SharpFont.Dependencies.

reaperrr committed:
Made sure spice blooms are drawn below harvesters all the time

Mailaender committed:
fix the Multiply blend mode

Mailaender committed:
Update README.md

add some additional new-lines to fix problems with local Markdown scripts

Robmaister committed:
Removed zlib from NSIS script

abcdefg30 committed:
Added BuildLimit: 1 to CYC2

pchote committed:
Use the correct (tile-based) radar color when updating the radar.

pchote committed:
Prevent resources from spawning on ramps.

Mailaender committed:
fix the weapon factory door opening animation

RoosterDragon committed:
Cache some trait lookups that occur frequently.

Mailaender committed:
add the flame animation

RoosterDragon committed:
Make IsDisabled a method on Actor.

This allows us to cache the disabled traits, which simplifies life for some callers since we relieve them of having to cache it, as well as improving perf for all IsDisabled calls.

RoosterDragon committed:
Cache results of TraitsImplementing calls.

If a class is caching the TraitsImplementing enumerable, instead cache the results of enumerating it to an array. The avoids having to enumerate the sequence each time it is needed.

pchote committed:
Rewrite yaml merging block removal.

Fixes #2922, #6818.

pchote committed:
Restore RallyPoint removal in soviet-01.

pchote committed:
Fix the MiniYaml unit tests.  Fixes #7999.

pchote committed:
Fix map yaml removals.

penev92 committed:
Move prone-related logic from DamageWarhead to TakeCover

Add DamageTypes to DamageWarhead and DamageTypeModifiers to TakeCover

RoosterDragon committed:
Prevent leaking of hardware cursors.

Added the lacking finialization machinery, ensured disposal under exceptional circumstances and also ensure the game calls dispose on old cursors before creating new ones.

RoosterDragon committed:
Retry hardware cursor creation.

On Windows, a spurious error sometimes occur when creating cursors. We retry a few times before before having to give up and fall back to software cursors.

pchote committed:
Support multiple inheritance for actor rules.

pchote committed:
Extract common inherits in ra defaults.yaml.

pchote committed:
Convert hidden [Desc] to a <summary>.

pchote committed:
Remove deprecated Buildable.Owner field.

obrakmann committed:
Clean up state properly when a Carryall's cargo has died

RoosterDragon committed:
Rename existing fields that are caching TraitsImplementing calls for consistency.

pchote committed:
Add a stub for upgrading map actors.

penev92 committed:
Add an upgrade rule for TakeCover

penev92 committed:
Add an upgrade rule for DamageWarheads

penev92 committed:
Add sandworm attack sound

abcdefg30 committed:
Fix a problem with hijackers in the survival maps

Mailaender committed:
deprecate RenderBuildingWarFactory

Mailaender committed:
add an upgrade rule

Mailaender committed:
remove trailing tabs

pchote committed:
Simplify and fix animation glitches.

sinf committed:
Added quick build feature

Mailaender committed:
fix d2k harvester sand overlay

pchote committed:
Upgrade mod rules.

Robmaister committed:
Fixed bug where the SDL2 directory wouldn't be deleted

Mailaender committed:
Update sounds-generic.yaml

switch building online/offline

Mailaender committed:
change voxel image on water terrain type for APC

pchote committed:
Split OrderButtonsChromeLogic into smaller classes.

pchote committed:
Split perf widgets into their own file.

Mailaender committed:
Update configure

Fixes #8036

pchote committed:
Fix colorpicker preview.

pchote committed:
Update widget mouseover when scrolling panels.

pchote committed:
Add an upgrade rule.

pchote committed:
Add an ExitDelay field to allow doors to open before exiting.

pchote committed:
Polish TS war factories.

abcdefg30 committed:
Use WithProductionDoorOverlay on fake weap

obrakmann committed:
Fix missing infantry in the legacy editor

The editor is looking for actor types that implement ILegacyEditorRender, but infantry units don't anymore since #7638.

This moves ILegacyEditorRender to the root of the Render* inheritance tree, so that all actor types have access to it.

Mailaender committed:
Update .travis.yml

define the mono version

pchote committed:
Polish TD weapons factory. Fixes #6651.

phrohdoh committed:
Add descriptions to CargoInfo members.

reaperrr committed:
Fix blue tiberium not appearing in legacy editor (TD)

RoosterDragon committed:
Ensure we never return a null string from LookupCountry.

phrohdoh committed:
Implement WithDecoration.

phrohdoh committed:
Implement WithDecorationDisguised.

phrohdoh committed:
Silence curl output for OS X native dependencies.

penev92 committed:
Introduce WithSpriteBody trait

Add upgrade rules

Add ISpriteBody

abcdefg30 committed:
Add an upgrade rule

phrohdoh committed:
Correct terminology and add readonly modifiers in SpriteEffect.

phrohdoh committed:
SpawnActorPower no longer forces removal of the spawned actor.

pchote committed:
Remove Moq dependency.

Mailaender committed:
use curl everywhere to run out of the box on Mac

Mailaender committed:
split render building/overlay and play sound for charge

to add the Tiberian Sun Nod Obelisk

tick slightly faster

tweak the local offsets

Mailaender committed:
clean the thirdparty directory, too

Mailaender committed:
remove non-existent DLL from list

Mailaender committed:
clean Windows dependencies as well

abcdefg30 committed:
Updated all maps with the utility

abcdefg30 committed:
Run the utility on all maps and polish stuff

pchote committed:
Configure native dependencies from the makefile.

pchote committed:
Move native deps script into thirdparty to avoid confusion.

Mailaender committed:
catch all nuget CLI downloads

Mailaender committed:
don't redownload cached files

Mailaender committed:
separate docs by namespace

Mailaender committed:
don't hardcode hidden mods

Mailaender committed:
fix trait documentation leaving out certain namespaces

Mailaender committed:
fix an overlap when the harvester exits to the north

penev92 committed:
Fix the Windows Make script

obrakmann committed:
Keep downloaded files in a separate directory

As a side effect, this re-enables building the Windows installer on Linux hosts.

obrakmann committed:
Make packaging less noisy, + random cleanups

obrakmann committed:
Adjust travis-ci caching

1. Removes the APT caching, since it's not available on the docker infrastructure.
2. Enables caching for the dependency download directory.

abcdefg30 committed:
Add selection boxes for dinos

abcdefg30 committed:
Add attack sounds for dinos

abcdefg30 committed:
Add funpark01

penev92 committed:
Rename ProvidesCustomPrerequisite to ProvidesPrerequisite

alexander-boll committed:
added mission 5 nod

Update nod05

penev92 committed:
Don't do trait lookups for Health on each processed order

alexander-boll committed:
added mission 6a nod

reaperrr committed:
Unharcode squad randomization, add minimum delay for creating attack force

penev92 committed:
Add tile BLOXTREE-345

pchote committed:
Only issue a MouseMove event if the scroll offset actually changed. Fixes #8098.

pchote committed:
Make sure that ScreenMap is initialized before other traits.

penev92 committed:
Cleanup unused DLL references

RoosterDragon committed:
Give objectives proper sizing.

This gives the container an actual size, which prevents the scroll panel logic skipping rendering in the belief that it is not visible.

pchote committed:
Remove Owner from ActorPreviews.

Mailaender committed:
also version the all helper mod

RoosterDragon committed:
Always access the SheetBuilder from the main thread in MapCache.

Not doing this can occasionally this can result in a cross thread call to the backing texture which leads to bad things. This stops trying to be clever regarding the thread safety invariants exposed by SheetBuilder and does things simply and safely: All updates happen on the main thread.

penev92 committed:
Add D2k faction descriptions to the lobby

reaperrr committed:
Significantly improved Necessary Illusions

- fixed all cliff tile errors
- fixed most noticable river errors
- fixed some noticable road errors
- fixed position of all tiberium trees (regression from #5198)

reaperrr committed:
Significantly improved Rogue States

- fixed most, possibly all cliff errors
- fixed many river, shore and road errors
- hid a few remaining shore errors behind trees/rocks
- fixed positions of all tiberium trees (1 cell down)
- replaced temperate tree groups with large desert tree
- tweaked position of some buildings at the border of the map so they're more visible (instead of half-hidden below shroud edge)

reaperrr committed:
Tweaked Break of Day and Chokepoint

Break of Day:
- fixed tiberium tree positions
- fixed a minor tile error near a river

Chokepoint:
- replaced temperate tree groups with large desert tree

reaperrr committed:
Converted rank and levelup art to cnc palette

reaperrr committed:
Split radar dish from Nod airfield and make it an idle overlay

reaperrr committed:
Add two previously unused voices to Commando voice set

reaperrr committed:
Fix WithIdleOverlay to start with first frame after make animation

phrohdoh committed:
Implement WithDecorationCarryable.

pchote committed:
Split PaletteReference into its own file.

pchote committed:
Rework player palette loading.

pchote committed:
Regenerate cached texture indices when the palette size changes.

pchote committed:
Don’t crash if IQBO isn’t present on an actor.

pchote committed:
Add ClickThrough property to ContainerWidget.

pchote committed:
Render terrain grid above actors, too.

pchote committed:
Force the first template variant in TTPW.

Mailaender committed:
filter actors with glitchy remap errors on certain tilesets

obrakmann committed:
Fix HotkeyEntryWidget keeping keyboard focus after receiving input

cjshmyr committed:
Give ally players a yellow hp bar for team health colors

abcdefg30 committed:
Add a low power overlay to fake structures

MorMund committed:
Fixed crash

Mailaender committed:
move ihptru up front

penev92 committed:
Add faction description tooltips to TD

abcdefg30 committed:
ArmorDrop has been removed

deniz1a committed:
Removes Spy's ability to disguise as Attack Dog and Giant Ant.

Commit:
Allow music download

Mailaender committed:
move generic Voxel render traits to Mods.Common

anderson2927 committed:
add a Windows dedicated server launch script

abcdefg30 committed:
Add a FormatTime function to UtilsGlobal.cs

abcdefg30 committed:
Remove trailing spaces from nod04b map.yaml

abcdefg30 committed:
Replace a bogus tab

deniz1a committed:
Fixes location of phase transport being revealed by attack line.

reaperrr committed:
Fixes the bug that selection voices are played twice when selecting groups
via 0-9 hotkeys.

Problem was that the voice was played for both KeyUp and KeyDown events.

Closes #8121.

Mailaender committed:
add skirmish stats

fixes #7728

pchote committed:
Overhaul save panel.

pchote committed:
Fix create map panel geometry.

Mailaender committed:
fix error CS0219:

The variable `tileStart' is assigned but its value is never used

Mailaender committed:
remove actors without palette glitches, but non-sensical

Mailaender committed:
allow 3rd-party mod DLLs on .NET > 4.0

Mailaender committed:
isometric spawn points

Mailaender committed:
isometric waypoints

deniz1a committed:
Reverts #8097.
It caused units not targeting enemies under fog even with satellite.

Baxxster committed:
Added possibility to use spaces in Server.Name

This way it's easier for others to set a servername with spaces in it by
default

Mailaender committed:
fix error CS0172

Mix-martes86 committed:
Replaced static shellmap with scripted one

penev92 commented:
Sorry, but while I'd welcome a more interesting shellmap, I don't think that a direct copy of RA's shellmap is a good idea:
 - The base layouts, while OK for RA, don't make much sense for D2k.
 - The map should highlight some D2k-specific features (carryalls, sandworms, starport delivery, Deviator tanks). Maybe have more emphasis on harvesting/protecting harvesters and the limited buildable space.

Mix-martes86 commented:
The idea for this shellmap is a year old (I just learned about Git basics and some more things, had no time until my recent unemployment), so obviously it has some holes.

My intention was to replace the, back then, completely empty and lifeless default shellmap on the release builds, at least temporarily, until an improved version came out, seeing how certain features can understandably be quite low in the priority queue. Or at least, that it could serve as sort of a template or tool for someone to make something nicer. And because I was long overdue since I spoke to people about it, I pushed it as is.

I thought about starport deliveries, but since I didn't know how to script them, I just left it as todo stuff. Carryalls are however in the map. Last release seems to automatically add them to harvesters without me doing anything. And sandworms, from what I've seen, can be ruled in as well without much hassle.

In any case, I'll see what can I do. I used a custom in-game editor (not done by me) back then, which considerably eased the tediousness of placing stuff or drawing the terrain, but I can't find the source.

penev92 commented:
The in-game map editor was actually merged last week, so you can find it in Main menu -> Extras -> Map editor.
Should make things much easier.
I am not convinced about the map/bases layout though.

Mailaender commented on mods/d2k/maps/shellmap/map.yaml:
```diff
@@ -33,102 +33,1571 @@ Players:
-			Maximum: 1
+		LoadWidgetAtGameStart:
+			Widget: MAINMENU
+		-StartGameNotification:
```
This triggers a lint rule error. https://travis-ci.org/OpenRA/OpenRA/builds/63595331#L1176

Mix-martes86 replied:
Yeah, noticed that when I made the request. I'll fix it in next commit.

Mailaender commented on mods/d2k/maps/shellmap/map.yaml:
```diff
@@ -33,102 +33,1571 @@ Players:
-			Minimum: 1
-			Maximum: 1
+		LoadWidgetAtGameStart:
+			Widget: MAINMENU
```
You don't need this anymore. We removed it during the preparation of the in-game map editor. https://github.com/OpenRA/OpenRA/pull/7217

Mix-martes86 replied:
Noted. On for the next commit.

Mailaender replied:
No, not the next commit. Use git amend and rebase to fix your previous commits.

Mix-martes86 committed:
Obsolete scripting and wormmanager within shellmap

Removed widget initialization and game notification calls, as well as
adding a WormSpawner.

Mix-martes86 commented:
I'm looking at this rebase thing, but I have absolutely no idea what I'm doing. I might break the commit history.

Mix-martes86 committed:
Merge remote-tracking branch 'origin/bleed' into bleed

Conflicts:
	OpenRA.Mods.RA/Traits/Render/RenderJammerCircle.cs

Mix-martes86 commented:
I give up, I don't want to break the origin repo by accident with this little experiment. This damned Git is the least user-friendly thing ever, I keep having to use a shell console to consolidate conflicts with already commited stuff, and I don't have the patience for that. I'll keep looking at improving the shellmap and upload it for someone to copy the code and do a pull request themselves if it raises any interest.

deniz1a commented:
You can try https://www.sourcetreeapp.com/ . It's user friendly. Also if you have problems with rebasing just clone a fresh copy of bleed and do your changes there...

Mix-martes86 commented:
That app is definitely friendlier looking, thanks. Will give it another try and see what happens.

x-a-n-a-x commented:
SmartGit is more stable and faster (in my opinion) but I agree with Deniz1a, don't give up.
 http://www.syntevo.com/smartgit/
